compiler/rustc_lint/src/dangling.rs

```diff
@@ -0,0 +1,196 @@
+use rustc_hir::{Expr, ExprKind, LangItem};
+use rustc_middle::ty::{self, Ty};
+use rustc_session::{declare_lint, declare_lint_pass};
+use rustc_span::symbol::sym;
+
+use crate::lints::InstantlyDangling;
+use crate::{LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `temporary_cstring_as_ptr` lint detects getting the inner pointer of
+    /// a temporary `CString`.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # #![allow(unused)]
+    /// # use std::ffi::CString;
+    /// let c_str = CString::new("foo").unwrap().as_ptr();
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// The inner pointer of a `CString` lives only as long as the `CString` it
+    /// points to. Getting the inner pointer of a *temporary* `CString` allows the `CString`
+    /// to be dropped at the end of the statement, as it is not being referenced as far as the
+    /// typesystem is concerned. This means outside of the statement the pointer will point to
+    /// freed memory, which causes undefined behavior if the pointer is later dereferenced.
+    pub TEMPORARY_CSTRING_AS_PTR,
+    Warn,
+    "detects getting the inner pointer of a temporary `CString`"
+}
+
+// FIXME: does not catch UnsafeCell::get
+// FIXME: does not catch getting a ref to a temporary and then converting it to a ptr
+declare_lint! {
+    /// The `instantly_dangling_pointer` lint detects getting a pointer to data
+    /// of a temporary that will immediately get dropped.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # #![allow(unused)]
+    /// # unsafe fn use_data(ptr: *const u8) {
+    /// #     dbg!(unsafe { ptr.read() });
+    /// # }
+    /// fn gather_and_use(bytes: impl Iterator<Item = u8>) {
+    ///     let x: *const u8 = bytes.collect::<Vec<u8>>().as_ptr();
+    ///     unsafe { use_data(x) }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Getting a pointer from a temporary value will not prolong its lifetime,
+    /// which means that the value can be dropped and the allocation freed
+    /// while the pointer still exists, making the pointer dangling.
+    /// This is not an error (as far as the type system is concerned)
+    /// but probably is not what the user intended either.
+    ///
+    /// If you need stronger guarantees, consider using references instead,
+    /// as they are statically verified by the borrow-checker to never dangle.
+    ///
+    /// Note: This lint does **not** get triggered by methods & functions
+    /// that intentionally produce dangling pointers, such as:
+    ///
+    /// - `core::ptr::dangling` & `core::ptr::dangling_mut`
+    /// - `core::ptr::NonNull::dangling`
+    /// - `std::alloc::Layout::dangling`
+    ///
+    pub INSTANTLY_DANGLING_POINTER,
```

Comment:
Lint names should be plural and more generally follow the [guidelines](https://rustc-dev-guide.rust-lang.org/diagnostics.html#lint-naming).

Maybe something like this? Which read well IMO: "allow dangling pointers from temporaries".
```suggestion
    pub DANGLING_POINTERS_FROM_TEMPORARIES,
```

Reply:
Done 08f9947579dc15fae743f7fe2a8b43046d3c5ce3